DS-4440 GDPR - Anonymize statistics feature

samuelcambien opened the pull request:
Fixes #7777

This PR adds a script to anonymise records in the Solr statistics core. The script will anonymise the IP values by rewriting (‘masking’) the last part. This mask is configurable, both for ipv4 and ipv6 addresses.

For IPv4 addresses, the last number will be replaced by the mask, defined by the configuration key ‘anonymise_statistics.ip_v4_mask’ which defaults to ‘254’.
For example, 109.74.16.171 is rewritten as 109.74.16.254

For IPv6 address, the last two numbers will be replaced by the mask, defined by the configuration key ‘anonymise_statistics.ip_v6_mask’ which defaults to ‘FFFF:FFFF’
For example, 2001:0db8:85a3:0000:0000:8a2e:0370:7334 is rewritten as 2001:0db8:85a3:0000:0000:8a2e:FFFF:FFFF

For each anonymised record, the DNS field is also replaced by “anonymised”.

The program only processes records older than 90 days. This period can be altered with the config ‘anonymise_statistics.time_limit’ (expressed in days).

The command ‘anonymise-statistics’ has been added to launcher.xml to launch the script.

The script takes an optional parameter ‘-s [sleep]’ (expressed in ms), which will make the Java thread sleep between the calls to Solr to reduce the load impact.

The Solr service commit mechanism is also optimised by adding multi-threading support. The script takes an optional parameter ‘-t [threads]’ to indicate how many threads the Solr service can use for this, if not given the thread count defaults to 2.

Besides the 'anonymise-statistics' script, which anonymises existing statistics record, the PR also allows for statistics records to be anonymised the moment they are created.
Enabling this feature can be done by setting the configuration parameter "anonymise_statistics.anonymise_on_log" to true.
When this configuration property is not set, the feature is disabled by default.

Finally, the PR also adds pagination support to SolrLoggerService

The work done on this PR is sponsored and led by the Humboldt-Universität zu Berlin.


This PR is related to https://jira.lyrasis.org/browse/DS-4440

Commit:
DS-4440 GDPR - Anonymize statistics feature

ssolim commented on dspace/config/launcher.xml:
```diff
@@ -380,4 +380,11 @@
+        <name>anonymize-statistics</name>
+        <description>Anonymize the ip values of the solr statistics</description>
+        <step>
+            <class>com.atmire.dspace.statistics.AnonymizeStatistics</class>
```
This looks like a typo to me. Shouldn't that be org.dspace.statistics.AnonymizeStatistics? ( pnb )

samuelcambien replied:
You are correct, fixed this!

Commit:
DS-4440 GDPR - Anonymize statistics feature - feedback

kshepherd reviewed (commented):
This PR is working well for me, I have tested various scenarios.
I also can see that it should not have any impact on sites that choose not to run the anonymize-statistics feature.

## Configuration defaults / properties

It would be good to add some configuration properties to **usage-statistics.cfg** as I can see five different properties with defaults used here (I needed to find the TIME_LIMIT one for my testing) but they're not included, even commented out, in any configuration files, and there is no documentation (yet) for this module.

In **AnonymizeStatistics** and **SolrServiceLoggerImpl**:
The 'anonymised' mask:
`private static final Object ANONYMISED = configurationService.getProperty("anonymise_statistics.dns_mask", "anonymised");`

The number of days before a record is anonymised
`calendar.add(DAY_OF_YEAR, -configurationService.getIntProperty("anonymise_statistics.time_limit", 90));`

In **SolrServiceLoggerImpl**:
Whether stats are anonymised on log:
`configurationService.getBooleanProperty("anonymise_statistics.anonymise_on_log", false)`

ipv4 mask:
`return ip.replaceFirst(IP_V4_REGEX, "$1" + configurationService.getProperty("anonymise_statistics.ip_v4_mask", "255"));`

ipv6 mask:
`return ip.replaceFirst(IP_V6_REGEX, "$1:" + configurationService.getProperty("anonymise_statistics.ip_v6_mask", "FFFF:FFFF"));`

Some default, commented out configuration properties in usage-statistics.cfg would be great, I think.

Also a reminder that we definitely need wiki doco for the launcher and the improvement in general, before 6.4 release

## Spelling (minor)
Not at all a blocker, but just noting that there is inconsistent usage of British vs American English with "anonymise" (used in config and in other parts of code) with "anonymize" (used in launcher and runner class, and other parts of code).
This isn't really a big deal but if it's easy enough to make it consistent it might help developers, especially those not familiar with the spelling variations, find the class or method or object they're looking for.

kshepherd commented:
hi @ssolim , now that the launcher is updated, if you're happy with the PR (i know you have it in production right?) could you change your review to "Approved"? Thanks!

samuelcambien committed:
DS-4440 GDPR - Anonymize statistics feature - feedback: default configuration & spelling

samuelcambien commented:
Hi,
Thanks for the feedback (and approval :) )
I've updated this PR:
* I followed the IDEA spelling suggestion to use 'anonymize' everywhere
* I've added commented out default configuration parameters to usage-statistics.cfg
* I've renamed 'anonymize_statistics.time_limit' to 'anonymize_statistics.time_threshold' because it makes more sense

samuelcambien committed:
DS-4440 GDPR - Anonymize statistics feature - feedback: default configuration & spelling bis

samuelcambien committed:
DS-4440 GDPR - Anonymize statistics feature - feedback: default configuration & spelling tris

AlexanderS commented on dspace/config/modules/usage-statistics.cfg:
```diff
@@ -40,4 +40,21 @@ usage-statistics.authorization.admin.workflow=true
+# anonymize_statistics.ip_v4_mask = 255"
+
+# Mask to replace the last two groups of an IPv6 address
+# anonymize_statistics.ip_v6_mask = FFFF:FFFF"
```
There are two superfluous quotes at the end of the two mask options.

AlexanderS commented on dspace/config/modules/usage-statistics.cfg:
```diff
@@ -40,4 +40,21 @@ usage-statistics.authorization.admin.workflow=true
+# Configuration parameters for anonymizing statistics
+
+# Anonymize statistics the moment they are created
+# anonymize_statistics.anonymize_on_log = false
```
The space before the config settings with the default values should be removed, so that there is a difference between comments and default config values.

AlexanderS reviewed (commented):
I just found two little issues with the default values in the config file.

kshepherd commented:
@samuelcambien i tried to push a small tidy up as per @AlexanderS 's comments but the atmire branch for this doesn't allow pushes from project maintainers.
Could you make this small tidy-up change? then we can get this merged :)
Would be great if the config could look like:

`##
# Configuration parameters for anonymizing statistics
##

# Anonymize statistics the moment they are created
#anonymize_statistics.anonymize_on_log = false

# Mask to replace the last group of an IPv4 address
#anonymize_statistics.ip_v4_mask = 255"

# Mask to replace the last two groups of an IPv6 address
#anonymize_statistics.ip_v6_mask = FFFF:FFFF"

# Mask to replace the DNS
#anonymize_statistics.dns_mask = anonymized

# Only anonymize statistics records older than this threshold (expressed in days)
#anonymize_statistics.time_threshold = 90`

AlexanderS commented:
Please remove the quotes at the end of the two masks, too:

> ```
> #anonymize_statistics.ip_v4_mask = 255"
>```

> ```
> #anonymize_statistics.ip_v6_mask = FFFF:FFFF"
> ```

samuelcambien committed:
DS-4440 GDPR - Anonymize statistics feature - feedback: default configuration

samuelcambien commented:
Hi, I've processed the latest feedback.

kshepherd commented:
With approvals here and from committers on the master port, I am merging this work, Thanks @samuelcambien !

MW3000 commented:
Would be nice to have this feature documented in: https://wiki.lyrasis.org/display/DSDOC8x/SOLR+Statistics
